File: hkgeocode.py
#!/usr/bin/env python3
"""Convert Hong Kong 1980 Grid (HK80) coordinates to HKGeoCode.

HKGeoCode (香港地理碼 / HKGCode) encodes a location in Hong Kong as up to
six Crockford Base32 characters at three nested resolutions:

    Level 1 (2 characters): 2 km × 2 km  district cell
    Level 2 (4 characters): 100 m × 100 m  neighbourhood cell
    Level 3 (6 characters): 5 m × 5 m  standard cell

The south-west origin is Easting 800000 m, Northing 800000 m. Coverage is
64 km east by 48 km north (32 × 24 large cells). Each pair of characters
is (easting index, northing index).

Definition and conversion steps:
    https://zh.wikipedia.org/zh-hk/香港地理碼
"""

from __future__ import annotations

import argparse
import math
import sys

# Crockford Base32 without I, L, O, U (to avoid confusion with 1 and 0).
CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"

ORIGIN_EASTING = 800_000.0
ORIGIN_NORTHING = 800_000.0

# (cell size in metres, easting index count, northing index count)
_LEVELS = (
    (2_000.0, 32, 24),  # large cell
    (100.0, 20, 20),  # medium cell
    (5.0, 20, 20),  # small cell
)

COVERAGE_EASTING = ORIGIN_EASTING + _LEVELS[0][0] * _LEVELS[0][1]  # 864000
COVERAGE_NORTHING = ORIGIN_NORTHING + _LEVELS[0][0] * _LEVELS[0][2]  # 848000

CELL_SIZE_M = {2: 2_000.0, 4: 100.0, 6: 5.0}

_LENGTHS = {2, 4, 6}


def coverage_grid_shape(cell_size: float = 100.0) -> tuple[int, int]:
    """Return (n_cols, n_rows_from_south) for the full HKGeoCode coverage."""
    n_cols = int(round((COVERAGE_EASTING - ORIGIN_EASTING) / cell_size))
    n_rows = int(round((COVERAGE_NORTHING - ORIGIN_NORTHING) / cell_size))
    return n_cols, n_rows


def hk80_to_grid_index(
    easting: float,
    northing: float,
    cell_size: float = 100.0,
) -> tuple[int, int]:
    """Return (column, row_from_south) of the cell containing the HK80 point.

    Column 0 is the westernmost cell; row 0 is the southernmost cell.
    """
    col = math.floor((easting - ORIGIN_EASTING) / cell_size)
    row = math.floor((northing - ORIGIN_NORTHING) / cell_size)
    n_cols, n_rows = coverage_grid_shape(cell_size)
    if not (0 <= col < n_cols and 0 <= row < n_rows):
        raise HKGeoCodeError(
            f"HK80 ({easting}, {northing}) is outside HKGeoCode coverage"
        )
    return int(col), int(row)


class HKGeoCodeError(ValueError):
    """Raised when coordinates or options are outside the HKGeoCode definition."""


def _encode_index(index: int) -> str:
    if not 0 <= index < len(CROCKFORD):
        raise HKGeoCodeError(f"index {index} is outside Crockford Base32 (0–31)")
    return CROCKFORD[index]


def hk80_to_hkgeocode(
    easting: float,
    northing: float,
    length: int = 6,
) -> str:
    """Return the HKGeoCode for an HK80 grid position.

    Args:
        easting: HK1980 Grid easting in metres.
        northing: HK1980 Grid northing in metres.
        length: 2, 4 or 6 characters (2 km, 100 m or 5 m precision).

    Returns:
        Upper-case HKGeoCode of the requested length.

    Raises:
        HKGeoCodeError: if ``length`` is invalid or the point is outside coverage.
    """
    if length not in _LENGTHS:
        raise HKGeoCodeError("length must be 2, 4 or 6")

    delta_e = easting - ORIGIN_EASTING
    delta_n = northing - ORIGIN_NORTHING

    if not (0.0 <= delta_e < COVERAGE_EASTING - ORIGIN_EASTING):
        raise HKGeoCodeError(
            f"easting {easting} is outside HKGeoCode coverage "
            f"[{ORIGIN_EASTING:g}, {COVERAGE_EASTING:g})"
        )
    if not (0.0 <= delta_n < COVERAGE_NORTHING - ORIGIN_NORTHING):
        raise HKGeoCodeError(
            f"northing {northing} is outside HKGeoCode coverage "
            f"[{ORIGIN_NORTHING:g}, {COVERAGE_NORTHING:g})"
        )

    chars: list[str] = []
    remainder_e = delta_e
    remainder_n = delta_n
    levels_needed = length // 2

    for size, x_count, y_count in _LEVELS[:levels_needed]:
        x_index = math.floor(remainder_e / size)
        y_index = math.floor(remainder_n / size)
        if not (0 <= x_index < x_count and 0 <= y_index < y_count):
            raise HKGeoCodeError(
                f"HK80 ({easting}, {northing}) is outside HKGeoCode coverage"
            )
        chars.append(_encode_index(x_index))
        chars.append(_encode_index(y_index))
        remainder_e -= x_index * size
        remainder_n -= y_index * size

    return "".join(chars)


def hkgeocode_cell_origin(code: str) -> tuple[float, float, float]:
    """Return (easting, northing, cell_size_m) of the cell's south-west corner."""
    code = code.strip().upper()
    if len(code) not in _LENGTHS or len(code) % 2:
        raise HKGeoCodeError("HKGeoCode must be 2, 4 or 6 characters")
    if any(ch not in CROCKFORD for ch in code):
        raise HKGeoCodeError(
            f"HKGeoCode {code!r} contains characters outside Crockford Base32"
        )

    easting = ORIGIN_EASTING
    northing = ORIGIN_NORTHING
    cell_size = _LEVELS[0][0]
    inverse = {ch: i for i, ch in enumerate(CROCKFORD)}

    for level, (size, x_count, y_count) in enumerate(_LEVELS):
        offset = level * 2
        if offset >= len(code):
            break
        x_index = inverse[code[offset]]
        y_index = inverse[code[offset + 1]]
        if x_index >= x_count or y_index >= y_count:
            raise HKGeoCodeError(
                f"HKGeoCode {code!r} has an out-of-range index at level {level + 1}"
            )
        easting += x_index * size
        northing += y_index * size
        cell_size = size

    return easting, northing, cell_size


def _self_test() -> int:
    """Check the Wikipedia / Lands Department worked examples."""
    cases = [
        # Coverage corners (large cells 00 and ZQ).
        (800_000.0, 800_000.0, 2, "00"),
        (862_000.0, 846_000.0, 2, "ZQ"),
        # Sharp Peak trig station 63: official HK80 sheet, Wikipedia WG73JD.
        (856_793.009, 832_366.405, 6, "WG73JD"),
        # South-west corner of the Clock Tower small cell from Wikipedia.
        (835_510.0, 817_185.0, 6, "H8FB2H"),
        (835_700.0, 817_200.0, 4, "H8HC"),
        (839_505.0, 817_015.0, 6, "K8FA13"),
        (828_420.0, 821_985.0, 6, "EA4K4H"),
        (802_000.0, 812_000.0, 2, "16"),
        (838_000.0, 820_000.0, 2, "KA"),
        (862_000.0, 844_000.0, 2, "ZP"),
        (844_000.0, 802_000.0, 2, "P1"),
    ]
    failed = 0
    for easting, northing, length, expected in cases:
        got = hk80_to_hkgeocode(easting, northing, length)
        status = "ok" if got == expected else "FAIL"
        if got != expected:
            failed += 1
        print(f"  {status}: E={easting} N={northing} -> {got} (expected {expected})")
    return failed


def _parse_args(argv: list[str] | None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Convert Hong Kong 1980 Grid easting/northing (metres) to HKGeoCode."
        )
    )
    parser.add_argument(
        "easting",
        nargs="?",
        type=float,
        help="HK80 easting in metres (positional alternative to -e)",
    )
    parser.add_argument(
        "northing",
        nargs="?",
        type=float,
        help="HK80 northing in metres (positional alternative to -n)",
    )
    parser.add_argument(
        "-e",
        "--easting",
        dest="easting_opt",
        type=float,
        metavar="METRES",
        help="HK80 easting in metres",
    )
    parser.add_argument(
        "-n",
        "--northing",
        dest="northing_opt",
        type=float,
        metavar="METRES",
        help="HK80 northing in metres",
    )
    parser.add_argument(
        "-l",
        "--length",
        type=int,
        choices=sorted(_LENGTHS),
        default=6,
        help="code length: 2 (2 km), 4 (100 m) or 6 (5 m). Default: 6",
    )
    parser.add_argument(
        "--self-test",
        action="store_true",
        help="run built-in checks against published examples and exit",
    )
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = _parse_args(argv)
    if args.self_test:
        failed = _self_test()
        return 1 if failed else 0

    easting = args.easting_opt if args.easting_opt is not None else args.easting
    northing = args.northing_opt if args.northing_opt is not None else args.northing
    if easting is None or northing is None:
        print(
            "error: easting and northing are required "
            "(positional EASTING NORTHING, or -e / -n)",
            file=sys.stderr,
        )
        return 2

    try:
        print(hk80_to_hkgeocode(easting, northing, args.length))
    except HKGeoCodeError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: js/hkgeocode.js
/**
 * HKGeoCode encode/decode for Hong Kong 1980 Grid (EPSG:2326).
 * Matches hkgeocode.py: Crockford Base32 pairs at 2 km, 100 m, and 5 m.
 */
(function (root) {
  "use strict";

  const CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
  const INVERSE = Object.fromEntries([...CROCKFORD].map((ch, i) => [ch, i]));
  const ORIGIN_EASTING = 800000;
  const ORIGIN_NORTHING = 800000;
  const LEVELS = [
    { size: 2000, xCount: 32, yCount: 24 },
    { size: 100, xCount: 20, yCount: 20 },
    { size: 5, xCount: 20, yCount: 20 },
  ];
  const LENGTHS = new Set([2, 4, 6]);
  const COVERAGE_EASTING = ORIGIN_EASTING + LEVELS[0].size * LEVELS[0].xCount;
  const COVERAGE_NORTHING = ORIGIN_NORTHING + LEVELS[0].size * LEVELS[0].yCount;

  class HKGeoCodeError extends Error {
    constructor(message) {
      super(message);
      this.name = "HKGeoCodeError";
    }
  }

  function encodeIndex(index) {
    if (index < 0 || index >= CROCKFORD.length) {
      throw new HKGeoCodeError(`index ${index} is outside Crockford Base32 (0–31)`);
    }
    return CROCKFORD[index];
  }

  function cellSizeForLength(length) {
    if (!LENGTHS.has(length)) {
      throw new HKGeoCodeError("length must be 2, 4 or 6");
    }
    return LEVELS[length / 2 - 1].size;
  }

  function coverageGridShape(cellSize) {
    return {
      nCols: Math.round((COVERAGE_EASTING - ORIGIN_EASTING) / cellSize),
      nRows: Math.round((COVERAGE_NORTHING - ORIGIN_NORTHING) / cellSize),
    };
  }

  function hk80ToHkgeocode(easting, northing, length = 6) {
    if (!LENGTHS.has(length)) {
      throw new HKGeoCodeError("length must be 2, 4 or 6");
    }
    const deltaE = easting - ORIGIN_EASTING;
    const deltaN = northing - ORIGIN_NORTHING;
    if (deltaE < 0 || deltaE >= COVERAGE_EASTING - ORIGIN_EASTING) {
      throw new HKGeoCodeError("easting is outside HKGeoCode coverage");
    }
    if (deltaN < 0 || deltaN >= COVERAGE_NORTHING - ORIGIN_NORTHING) {
      throw new HKGeoCodeError("northing is outside HKGeoCode coverage");
    }
    const chars = [];
    let remE = deltaE;
    let remN = deltaN;
    const levelsNeeded = length / 2;
    for (let i = 0; i < levelsNeeded; i += 1) {
      const { size, xCount, yCount } = LEVELS[i];
      const xIndex = Math.floor(remE / size);
      const yIndex = Math.floor(remN / size);
      if (xIndex < 0 || yIndex < 0 || xIndex >= xCount || yIndex >= yCount) {
        throw new HKGeoCodeError("coordinates are outside HKGeoCode coverage");
      }
      chars.push(encodeIndex(xIndex), encodeIndex(yIndex));
      remE -= xIndex * size;
      remN -= yIndex * size;
    }
    return chars.join("");
  }

  function normalizeCode(code) {
    const normalized = String(code || "").trim().toUpperCase();
    if (!LENGTHS.has(normalized.length)) {
      throw new HKGeoCodeError("HKGeoCode must be 2, 4 or 6 characters");
    }
    for (const ch of normalized) {
      if (!(ch in INVERSE)) {
        throw new HKGeoCodeError(
          `HKGeoCode ${normalized} contains characters outside Crockford Base32`
        );
      }
    }
    return normalized;
  }

  function hkgeocodeCellOrigin(code) {
    const normalized = normalizeCode(code);
    let easting = ORIGIN_EASTING;
    let northing = ORIGIN_NORTHING;
    let cellSize = LEVELS[0].size;
    for (let level = 0; level < LEVELS.length; level += 1) {
      const offset = level * 2;
      if (offset >= normalized.length) {
        break;
      }
      const { size, xCount, yCount } = LEVELS[level];
      const xIndex = INVERSE[normalized[offset]];
      const yIndex = INVERSE[normalized[offset + 1]];
      if (xIndex >= xCount || yIndex >= yCount) {
        throw new HKGeoCodeError(
          `HKGeoCode ${normalized} has an out-of-range index at level ${level + 1}`
        );
      }
      easting += xIndex * size;
      northing += yIndex * size;
      cellSize = size;
    }
    return { easting, northing, cellSize };
  }

  function parentCode(code) {
    const normalized = normalizeCode(code);
    if (normalized.length === 2) {
      return null;
    }
    return normalized.slice(0, normalized.length - 2);
  }

  function neighbourCodes(code) {
    const normalized = normalizeCode(code);
    const { easting, northing, cellSize } = hkgeocodeCellOrigin(normalized);
    const half = cellSize / 2;
    const neighbours = [];
    for (const de of [-1, 0, 1]) {
      for (const dn of [-1, 0, 1]) {
        if (de === 0 && dn === 0) {
          continue;
        }
        try {
          neighbours.push(
            hk80ToHkgeocode(
              easting + half + de * cellSize,
              northing + half + dn * cellSize,
              normalized.length
            )
          );
        } catch (err) {
          if (!(err instanceof HKGeoCodeError)) {
            throw err;
          }
        }
      }
    }
    return neighbours;
  }

  function childCodes(code) {
    const normalized = normalizeCode(code);
    if (normalized.length === 6) {
      return [];
    }
    const nextLength = normalized.length + 2;
    const { easting, northing, cellSize: parentSize } = hkgeocodeCellOrigin(normalized);
    const childSize = cellSizeForLength(nextLength);
    const n = Math.round(parentSize / childSize);
    const children = [];
    for (let i = 0; i < n; i += 1) {
      for (let j = 0; j < n; j += 1) {
        children.push(
          hk80ToHkgeocode(
            easting + (i + 0.5) * childSize,
            northing + (j + 0.5) * childSize,
            nextLength
          )
        );
      }
    }
    return children;
  }

  function cellIntersects(code, west, south, east, north) {
    const { easting, northing, cellSize } = hkgeocodeCellOrigin(code);
    return (
      easting < east &&
      easting + cellSize > west &&
      northing < north &&
      northing + cellSize > south
    );
  }

  function codesInBounds(west, south, east, north, length) {
    const size = cellSizeForLength(length);
    const { nCols, nRows } = coverageGridShape(size);
    const col0 = Math.max(0, Math.floor((west - ORIGIN_EASTING) / size));
    const row0 = Math.max(0, Math.floor((south - ORIGIN_NORTHING) / size));
    const col1 = Math.min(nCols - 1, Math.floor((east - ORIGIN_EASTING - 1e-9) / size));
    const row1 = Math.min(nRows - 1, Math.floor((north - ORIGIN_NORTHING - 1e-9) / size));
    if (col1 < col0 || row1 < row0) {
      return [];
    }
    const codes = [];
    for (let row = row0; row <= row1; row += 1) {
      for (let col = col0; col <= col1; col += 1) {
        const e = ORIGIN_EASTING + (col + 0.5) * size;
        const n = ORIGIN_NORTHING + (row + 0.5) * size;
        try {
          codes.push(hk80ToHkgeocode(e, n, length));
        } catch (err) {
          if (!(err instanceof HKGeoCodeError)) {
            throw err;
          }
        }
      }
    }
    return codes;
  }

  function levelLabel(length) {
    if (length === 2) {
      return "District (2 km)";
    }
    if (length === 4) {
      return "Neighbourhood (100 m)";
    }
    return "Standard (5 m)";
  }

  const api = {
    CROCKFORD,
    ORIGIN_EASTING,
    ORIGIN_NORTHING,
    COVERAGE_EASTING,
    COVERAGE_NORTHING,
    LEVELS,
    HKGeoCodeError,
    cellSizeForLength,
    coverageGridShape,
    hk80ToHkgeocode,
    hkgeocodeCellOrigin,
    parentCode,
    childCodes,
    neighbourCodes,
    cellIntersects,
    codesInBounds,
    levelLabel,
    normalizeCode,
  };

  root.HKGeoCode = api;
})(typeof window !== "undefined" ? window : globalThis);
